lib/src/timing.dart

```diff
@@ -0,0 +1,371 @@
+}
+
+/// Implementation of [SyncTimeTracker] that can handle async actions
+///
+/// Async actions returning [Future] will be tracked as single sync time span
+/// from the beginning of execution till completion of future
+///
+/// Use [AsyncTimeTracker.simple] factory to get an instance.
+class SimpleAsyncTimeTrackerImpl extends TimeSliceGroup
+    implements AsyncTimeTracker {
+  @override
+  R track<T, R extends FutureOr<T>>(R Function() action) {
+    if (isStarted) {
+      throw StateError('Can not be tracked twice');
+    }
+    R result;
+    var tracker = SyncTimeTrackerImpl();
+    slices.add(tracker);
+    tracker.start();
+    try {
+      result = action();
+    } catch (_) {
+      tracker.stop();
+      rethrow;
+    }
+    if (result is Future<T>) {
+      return result.whenComplete(tracker.stop) as R;
+    } else {
+      tracker.stop();
+      return result;
+    }
+  }
+
+  @override
+  bool get isStarted => slices.isNotEmpty;
+
+  @override
+  bool get isTracking =>
+      slices.isNotEmpty && (slices.first as SyncTimeTracker).isTracking;
+
+  @override
+  bool get isFinished =>
+      slices.isNotEmpty && (slices.first as SyncTimeTracker).isFinished;
+}
+
+/// No-op implementation of [SyncTimeTracker] that does nothing.
+///
+/// Use [SyncTimeTracker.noOp] factory to get an instance.
+class NoOpSyncTimeTracker implements SyncTimeTracker {
+  static final sharedInstance = NoOpSyncTimeTracker();
+
+  @override
+  Duration get duration =>
+      throw UnsupportedError('Unsupported in no-op implementation');
+
+  @override
+  DateTime get startTime =>
+      throw UnsupportedError('Unsupported in no-op implementation');
+
+  @override
+  DateTime get stopTime =>
+      throw UnsupportedError('Unsupported in no-op implementation');
+
+  @override
+  bool get isStarted =>
+      throw UnsupportedError('Unsupported in no-op implementation');
+
+  @override
+  bool get isTracking =>
+      throw UnsupportedError('Unsupported in no-op implementation');
+
+  @override
+  bool get isFinished =>
+      throw UnsupportedError('Unsupported in no-op implementation');
+
+  @override
+  R track<T, R extends FutureOr<T>>(R Function() action) => action();
+}
+
+abstract class AsyncTimeTracker implements TimeSliceGroup, TimeTracker {
+  factory AsyncTimeTracker({bool trackNested = true}) =>
+      AsyncTimeTrackerImpl(trackNested);
+
+  factory AsyncTimeTracker.simple() => SimpleAsyncTimeTrackerImpl();
+
+  factory AsyncTimeTracker.noOp() => NoOpAsyncTimeTracker.sharedInstance;
+}
+
+/// Implementation of a real [AsyncTimeTracker].
+///
+/// Use [AsyncTimeTracker] factory to get an instance.
+class AsyncTimeTrackerImpl extends TimeSliceGroup implements AsyncTimeTracker {
+  final bool trackNested;
+
+  AsyncTimeTrackerImpl(this.trackNested);
+
+  T _trackSyncSlice<T>(ZoneDelegate parent, Zone zone, T Function() action) {
+    // Ignore dangling runs after tracker completes
+    if (isFinished) {
+      return action();
```

Comment:
I don't really know what it mean. In "complex" test case, when I've added completer instead of relaying on timer/microtask priorities this dangling run call has appeared. It is called after future returned by original action passed to track is completed. So even if caller of `track` awaits it result, it will be called after awaiting and mess up with result.